apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/RpcRetryConsumer.cs

```diff
@@ -0,0 +1,214 @@
+        /// action based on retry options.
+        /// </summary>
+        private readonly Func<Action, RetryOptions, ISequentialThreadingTimer> _timerFactory;
+
+        /// <summary>The timer running in the background to retry failed requests.</summary>
+        private readonly ISequentialThreadingTimer _timer;
+
+        /// <summary>A semaphore to protect the buffer.</summary>
+        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1);
+
+        /// <summary>The consumer to send items to.</summary>
+        internal readonly IConsumer<T> _consumer;
+
+        /// <summary>
+        /// The buffered items. This is not readonly as it is replaced when the buffer is successfully send to
+        /// the base consumer.
+        /// </summary>
+        private List<T> _buffer = new List<T>();
+
+        /// <summary>The size of the buffer.</summary>
+        private int _size = 0;
+
+        /// <summary>The number of sequential unsuccessful retries.</summary>
+        private int _retries = 0;        
+
+        internal RpcRetryConsumer(IConsumer<T> consumer, RetryOptions options, 
+            Func<T, int> sizer, Func<Action, RetryOptions, ISequentialThreadingTimer> timerFactory)
+        {
+            _consumer = GaxPreconditions.CheckNotNull(consumer, nameof(consumer));
+            _options = GaxPreconditions.CheckNotNull(options, nameof(options));
+            _sizer = GaxPreconditions.CheckNotNull(sizer, nameof(sizer));
+            _timerFactory = GaxPreconditions.CheckNotNull(timerFactory, nameof(timerFactory));
+
+            // Start the background timer that will retry failed requests.
+            _timer = _timerFactory(FlushBuffer, _options);
+        }
+
+        /// <inheritdoc />
+        public void Receive(IEnumerable<T> items)
+        {
+            GaxPreconditions.CheckNotNull(items, nameof(items));
+            try
+            {
+                _consumer.Receive(items);
+            }
+            catch (RpcException e)
+            {
+                HandleException(items, e);
+            }
+        }
+
+        /// <inheritdoc />
+        public Task ReceiveAsync(IEnumerable<T> items, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            GaxPreconditions.CheckNotNull(items, nameof(items));
+            try
+            {
+                return _consumer.ReceiveAsync(items, cancellationToken);
+            }
+            catch (RpcException e)
+            {
+                HandleException(items, e);
+                return CommonUtils.CompletedTask;
+            }
+        }
+
+        /// <inheritdoc />
+        public void Dispose() => _timer?.Dispose();
+
+        /// <summary>
+        /// Handles a <see cref="RpcException"/> that occurred during a <see cref="Receive(IEnumerable{T})"/> call.
+        /// </summary>
+        private void HandleException(IEnumerable<T> items, RpcException e)
```

Comment:
Maybe have this return a bool and use it in an exception filter where this is needed. Return false when you want to propagate the exception so you don't lose stack trace details.